Make row number generation optional in RowNumber operator

czentgr opened the pull request:
Changes in the Presto coordinator, made with PR
https://github.com/prestodb/presto/pull/19835 allows the row number generation to be skipped for a partial RowNumber node that feeds the Exchange operator. It reduces the amount of data sent.

The implementation is similar to the TopNRowNumber node and operator capability to skip row generation.

A subsequent PR is needed for Prestissimo to make use of the new capability when converting the Presto plan into the Velox plan.

This resolves the end-to-end test issue discussed in issue https://github.com/prestodb/presto/issues/20227.

czentgr commented:
velox_exec_test results:

```
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test "--gtest_filter=RowNumberTest.*"
Note: Google Test filter = RowNumberTest.*
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from RowNumberTest
[ RUN      ] RowNumberTest.basic
[       OK ] RowNumberTest.basic (38 ms)
[ RUN      ] RowNumberTest.noPartitionKeys
[       OK ] RowNumberTest.noPartitionKeys (115 ms)
[ RUN      ] RowNumberTest.largeInput
[       OK ] RowNumberTest.largeInput (2531 ms)
[----------] 3 tests from RowNumberTest (2685 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (2712 ms total)
[  PASSED  ] 3 tests.
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test "--gtest_filter=PlanNodeSerdeTest.rowNumber"
Note: Google Test filter = PlanNodeSerdeTest.rowNumber
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from PlanNodeSerdeTest
[ RUN      ] PlanNodeSerdeTest.rowNumber
[       OK ] PlanNodeSerdeTest.rowNumber (33 ms)
[----------] 1 test from PlanNodeSerdeTest (34 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (34 ms total)
[  PASSED  ] 1 test.
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test "--gtest_filter=PlanNodeToStringTest.rowNumber"
Note: Google Test filter = PlanNodeToStringTest.rowNumber
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from PlanNodeToStringTest
[ RUN      ] PlanNodeToStringTest.rowNumber
[       OK ] PlanNodeToStringTest.rowNumber (35 ms)
[----------] 1 test from PlanNodeToStringTest (35 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (35 ms total)
[  PASSED  ] 1 test.
```

aditi-pandit commented on velox/exec/tests/utils/PlanBuilder.cpp:
```diff
@@ -1412,9 +1412,18 @@ PlanBuilder& PlanBuilder::window(
     const std::vector<std::string>& partitionKeys,
-    std::optional<int32_t> limit) {
+    std::optional<int32_t> limit,
+    const bool generateRowNumber) {
```
Any reason to not use std::optional for rowNumberColumnName ? That's how the TopNRowNumberNode implements also. And the column name wouldn't be hard-coded.

czentgr replied:
This is the test PlanBuilder for RowNumber node.
The corresponding signature for TopNRow is
```
PlanBuilder& PlanBuilder::topNRowNumber(
    const std::vector<std::string>& partitionKeys,
    const std::vector<std::string>& sortingKeys,
    int32_t limit,
    bool generateRowNumber) {
```
It is the same approach using a boolean to indicate whether or not the node constructor gets passed the column name via the optional.
The actual RowNumber node constructor does use an optional just like the TopNRowNumber node.

Am I missing something?

aditi-pandit replied:
My bad... I misread this. For PlanBuilder that is fine.

czentgr commented:
Failure in MemoryAllocatorTests. While the operator would use less memory (does not have to allocate the additional bigint for row numbers) it needs to be explicitly enabled/set. Therefore, I don't think it is related but will take a look.

Failure:
```
44: [ RUN      ] MemoryAllocatorTests/MemoryAllocatorTest.contiguousAllocation/0
38: ../../velox/common/memory/tests/MemoryAllocatorTest.cpp:154: Failure
38: Expected: (expected) <= (current.vsize), actual: 66549 vs 60804
38: ../../velox/common/memory/tests/MemoryAllocatorTest.cpp:157: Failure
38: Expected: (expected) <= (current.rss), actual: 17712 vs 12337
38: [  FAILED  ] MemoryAllocatorTests/MemoryAllocatorTest.allocContiguousVsize/0, where GetParam() = false (42 ms)
```

czentgr commented:
Running the Memory test locally succeeds. It is probably flaky.

```
czentgr@Christians-MacBook-Pro tests % ./velox_memory_test "--gtest_filter=*/MemoryAllocatorTest.allocContiguousVsize/*"
Note: Google Test filter = */MemoryAllocatorTest.allocContiguousVsize/*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from MemoryAllocatorTests/MemoryAllocatorTest
[ RUN      ] MemoryAllocatorTests/MemoryAllocatorTest.allocContiguousVsize/0
[       OK ] MemoryAllocatorTests/MemoryAllocatorTest.allocContiguousVsize/0 (12 ms)
[ RUN      ] MemoryAllocatorTests/MemoryAllocatorTest.allocContiguousVsize/1
[       OK ] MemoryAllocatorTests/MemoryAllocatorTest.allocContiguousVsize/1 (11 ms)
[----------] 2 tests from MemoryAllocatorTests/MemoryAllocatorTest (24 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (24 ms total)
[  PASSED  ] 2 tests.
```

aditi-pandit commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -22,17 +22,32 @@ class RowNumberTest : public OperatorTestBase {};
   auto data = makeRowVector({
-      makeFlatVector<int64_t>({1, 2, 1, 2, 1, 2, 1}),
-      makeFlatVector<int64_t>({1, 2, 3, 4, 5, 6, 7}),
+      makeFlatVector<int64_t>({1, 2, 1, 2, 1, 2, 1}), // c0, paritioning key.
```
Please remove these comments. The column names c0 and c1 are implicit and the usage of the partitioning and data depends on the individual tests.

aditi-pandit commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -22,17 +22,32 @@ class RowNumberTest : public OperatorTestBase {};
   auto plan = PlanBuilder().values({data}).rowNumber({"c0"}).planNode();
   assertQuery(plan, "SELECT *, row_number() over (partition by c0) FROM tmp");
 
+  plan = PlanBuilder().values({data}).rowNumber({"c0"}).planNode();
```
This test seems redundant. It can be removed.

aditi-pandit commented on velox/exec/tests/RowNumberTest.cpp:
```diff
@@ -55,11 +80,25 @@ TEST_F(RowNumberTest, noPartitionKeys) {
   auto plan = PlanBuilder().values({data, data}).rowNumber({}).planNode();
   assertQuery(plan, "SELECT *, row_number() over () FROM tmp");
 
+  // No limit, emit row numbers.
```
This test seems unnecessary. Please remove it.

aditi-pandit reviewed (commented):
Thanks @czentgr. Just have some minor comments.

czentgr commented:
Latest test runs after addressing review comments.

```
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test "--gtest_filter=RowNumberTest.*"
Note: Google Test filter = RowNumberTest.*
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from RowNumberTest
[ RUN      ] RowNumberTest.basic
[       OK ] RowNumberTest.basic (42 ms)
[ RUN      ] RowNumberTest.noPartitionKeys
[       OK ] RowNumberTest.noPartitionKeys (102 ms)
[ RUN      ] RowNumberTest.largeInput
[       OK ] RowNumberTest.largeInput (2292 ms)
[----------] 3 tests from RowNumberTest (2437 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (2465 ms total)
[  PASSED  ] 3 tests.
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test "--gtest_filter=PlanNodeToStringTest.rowNumber"
Note: Google Test filter = PlanNodeToStringTest.rowNumber
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from PlanNodeToStringTest
[ RUN      ] PlanNodeToStringTest.rowNumber
[       OK ] PlanNodeToStringTest.rowNumber (24 ms)
[----------] 1 test from PlanNodeToStringTest (24 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (24 ms total)
[  PASSED  ] 1 test.
czentgr@Christians-MacBook-Pro tests % ./velox_exec_test "--gtest_filter=PlanNodeSerdeTest.rowNumber"
Note: Google Test filter = PlanNodeSerdeTest.rowNumber
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from PlanNodeSerdeTest
[ RUN      ] PlanNodeSerdeTest.rowNumber
[       OK ] PlanNodeSerdeTest.rowNumber (23 ms)
[----------] 1 test from PlanNodeSerdeTest (23 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (23 ms total)
[  PASSED  ] 1 test.
```

czentgr commented:
@aditi-pandit Thanks for you review! I've addressed your review comments.
@mbasmanova Please review when you get a chance. Thank you!

aditi-pandit commented on velox/exec/RowNumber.cpp:
```diff
@@ -155,10 +164,12 @@ RowVectorPtr RowNumber::getOutputForSinglePartition() {
-  for (auto i = 0; i < numOutput; ++i) {
-    rowNumbers.set(i, ++numTotalInput_);
+  if (generateRowNumber_) {
+    FlatVector<int64_t>* rowNumbers =
```
You can use "auto" since the type of rowNumbers is explicitly written in the casting.

czentgr replied:
Yes. I had the same thought after the fact that I could use that.

aditi-pandit commented on velox/exec/RowNumber.cpp:
```diff
@@ -119,7 +126,9 @@ RowVectorPtr RowNumber::getOutput() {
     }
 
-    rowNumbers.set(i, rowNumber);
+    if (rowNumbers) {
```
Nit : You can use generateRowNumber_ like other usages.

aditi-pandit reviewed (approved):
Thanks @czentgr. Few more nits.

mbasmanova commented on velox/exec/RowNumber.cpp:
```diff
@@ -74,14 +77,14 @@ void RowNumber::addInput(RowVectorPtr input) {
 }
 
-FlatVector<int64_t>& RowNumber::getOrCreateRowNumberVector(vector_size_t size) {
+VectorPtr RowNumber::getOrCreateRowNumberVector(vector_size_t size) {
```
Why is this change?

czentgr replied:
At first I changed to the type to a pointer type and after review to the type you commented on. My original idea was so it can return a value that can be easily assigned (reference types need to be initialized and this wouldn't have worked).
But, the original function returns an l-value and I can take the address of that and assign to a pointer type variable. I reverted the change and adjusted accordingly.

mbasmanova reviewed (commented):
@czentgr Looks good % one question.

mbasmanova commented:
@czentgr Would you follow up with update to the documentation: velox/docs/develop/operators.rst ?

czentgr commented:
> @czentgr Would you follow up with update to the documentation: velox/docs/develop/operators.rst ?

@mbasmanova A quick review shows for `TopNRowNumberNode` the fact that the row number generation is optional is not mentioned/discussed. I suppose the doc needs an update for both RowNumber and TopNRowNumberNode operators regarding this.

I will make a doc update. Should I do it in this PR or another PR?

czentgr commented:
Looks like the linter is complaining but I can't see what it is.

mbasmanova commented:
> I will make a doc update. Should I do it in this PR or another PR?

Thanks. Let's make it in a separate PR.

czentgr commented:
@mbasmanova What about the linter warning that I have no insight into?

mbasmanova commented:
> @mbasmanova What about the linter warning that I have no insight into?

Looks like a false positive. I'll let you know if anything will be blocking the merge.
